django/contrib/admin/widgets.py

```diff
@@ -402,3 +404,88 @@ def __init__(self, attrs=None):
         if attrs is not None:
             final_attrs.update(attrs)
         super(AdminCommaSeparatedIntegerFieldWidget, self).__init__(attrs=final_attrs)
+
+
+class AdminForeignKeyWidget(forms.Select):
+    """
+    ForeignKey widget that loads options via AJAX to avoid rendering large Querysets.
+
+    This class is responsible for rendering the necessary
+    data attributes for select2 as well as adding the static
+    form media.
+    """
+
+    url_name = 'admin:fk_json'
+
+    def __init__(self, model_admin, attrs=None, using=None):
+        self.model_admin = model_admin
+        self.db = using
+        super(AdminForeignKeyWidget, self).__init__(attrs)
+
+    def get_url(self):
+        """Return URL from instance or by reversing :attr:`.url_name`."""
+        return reverse(self.url_name)
+
+    def render(self, name, value, attrs=None):
+        if value is None:
+            value = ''
+        final_attrs = self.build_attrs(attrs, name=name)
+        final_attrs['data-field_identifier'] = self.get_field_identifier(name)
+        output = [format_html('<select{}>', flatatt(final_attrs))]
+        options = self.render_options([value])
+        if options:
+            output.append(options)
+        output.append('</select>')
+        return mark_safe('\n'.join(output))
+
+    def build_attrs(self, extra_attrs=None, **kwargs):
+        """Set select2's AJAX attributes."""
+        attrs = super(AdminForeignKeyWidget, self).build_attrs(extra_attrs=extra_attrs, **kwargs)
+        attrs.setdefault('class', '')
+        attrs.setdefault('data-minimum-input-length', 0)
+        attrs.update({
+            'data-ajax--url': self.get_url(),
+            'data-ajax--cache': "true",
+            'data-ajax--type': "GET",
+            'data-allow-clear': json.dumps(not self.is_required),
+            'data-placeholder': '',  # Needs to be empty to allow clearing of the input
+            'class': attrs['class'] + 'admin-foreignkey'
+        })
+        return attrs
+
+    def get_field_identifier(self, name):
+        model = self.model_admin.model
+        return "{app_label}.{model_name}.{field_name}".format(
+            app_label=model._meta.app_label,
+            model_name=model.__name__,
+            field_name=name,
```

Comment:
Unfortunately this doesn't work well for formsets. I get `order.OrdersProducts.ordersproducts_set-15-product`.